File: app/src/main/java/com/singword/app/domain/VocabMatcher.kt
package com.singword.app.domain

import com.singword.app.data.local.wordbook.WordEntry

object VocabMatcher {

    /**
     * Match tokenized words against loaded wordbooks.
     * @param tokens  Set of lowercase words from lyrics
     * @param wordbooks  Map<lowercase_word, Pair<WordEntry, source_label>>
     * @return List of matched words, sorted alphabetically
     */
    fun match(
        tokens: Set<String>,
        wordbooks: Map<String, Pair<WordEntry, String>>
    ): List<MatchedWord> {
        return tokens
            .mapNotNull { token ->
                wordbooks[token]?.let { (entry, source) ->
                    MatchedWord(
                        word = entry.word,
                        pos = entry.pos,
                        def = entry.def,
                        source = source
                    )
                }
            }
            .sortedBy { it.word }
    }
}

File: ios/singword/Domain/VocabMatcher.swift
import Foundation

enum VocabMatcher {
    static func match(
        tokens: Set<String>,
        wordbooks: [String: (WordEntry, String)]
    ) -> [MatchedWord] {
        tokens
            .compactMap { token in
                guard let (entry, source) = wordbooks[token] else {
                    return nil
                }
                return MatchedWord(
                    word: entry.word,
                    pos: entry.pos,
                    definition: entry.definition,
                    source: source
                )
            }
            .sorted { $0.word < $1.word }
    }
}
